[Cocoa] Use AVCaptureDeviceDiscoverySession for camera discovery

eric-carlson opened the pull request:
#### 460662e8590c4a2e99a90a8a97b56ed9cb92016f
<pre>
[Cocoa] Use AVCaptureDeviceDiscoverySession for camera discovery
<a href="https://bugs.webkit.org/show_bug.cgi?id=248054">https://bugs.webkit.org/show_bug.cgi?id=248054</a>
rdar://101919809

Reviewed by Youenn Fablet.

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:

* Source/WebCore/platform/mediastream/mac/AVCaptureDeviceManager.mm:
(WebCore::currentCameras): Use AVCaptureDeviceDiscoverySession.
(WebCore::AVCaptureDeviceManager::updateCachedAVCaptureDevices): Call currentCameras.
(WebCore::AVCaptureDeviceManager::retrieveCaptureDevices): Ditto.
(WebCore::AVCaptureDeviceManager::registerForDeviceNotifications): Key-value observe
AVCaptureDeviceDiscoverySession.devices.
(WebCore::AVCaptureDeviceManager::~AVCaptureDeviceManager): Stop listening.

Canonical link: <a href="https://commits.webkit.org/256849@main">https://commits.webkit.org/256849@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65bd753c083c25121c745faa0f07c936b66d4dfe

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96979 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6246 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30092 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106507 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166791 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100953 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6474 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34981 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89370 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103204 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102649 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4866 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83595 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31880 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86709 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88577 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74771 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/280 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/264 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4738 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5056 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1496 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40785 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

youennf reviewed (approved):
Is this API also working properly in older macOSes?

eric-carlson committed:
[Cocoa] Use AVCaptureDeviceDiscoverySession for camera discovery
https://bugs.webkit.org/show_bug.cgi?id=248054
rdar://101919809

Reviewed by Youenn Fablet.

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:

* Source/WebCore/platform/mediastream/mac/AVCaptureDeviceManager.mm:
(WebCore::currentCameras): Use AVCaptureDeviceDiscoverySession.
(WebCore::AVCaptureDeviceManager::updateCachedAVCaptureDevices): Call currentCameras.
(WebCore::AVCaptureDeviceManager::retrieveCaptureDevices): Ditto.
(WebCore::AVCaptureDeviceManager::registerForDeviceNotifications): Key-value observe
AVCaptureDeviceDiscoverySession.devices.
(WebCore::AVCaptureDeviceManager::~AVCaptureDeviceManager): Stop listening.

Canonical link: https://commits.webkit.org/256849@main
